src/scatter.jl

```diff
@@ -9,17 +9,39 @@
 #     - ∇scatter_src!
 #
 
-function _check_dims(Ndst, Nsrc, N, Nidx)
-    @assert Ndst - N == Nsrc - Nidx "Incompatible input shapes of (dst, src, idx) = ($Ndst, $Nsrc, $Nidx)."
-    dims = Ndst - N
-    if dims < 0
-        throw(ArgumentError("dims must be non-negative but got dims=$dims."))
-    end
+typelength(::Type{<:Number}) = 1
+typelength(::Type{<:NTuple{M}}) where M = M
+typelength(::Type{CartesianIndex{M}}) where M = M
+
+function _check_dims(X::AbstractArray{Tx,Nx}, 
+                     Y::AbstractArray{Ty,Ny},
+                     idx::AbstractArray{Tidx,Nidx}) where
+                     {Tx,Ty,Tidx<:IntOrIntTuple,Nx,Ny,Nidx}
+    M = typelength(Tidx)
+    dims = _check_dims(Nx, Ny, M, Nidx)
+    size(X)[1:dims] == size(Y)[1:dims] || throw(ArgumentError("Incompatible input shapes."))
+    size(Y)[dims+1:end] == size(idx) || throw(ArgumentError("Incompatible input shapes."))
     return dims
 end
 
-typelength(::Type{<:Number}) = 1
-typelength(::Type{<:NTuple{M}}) where M = M
+function _check_dims(X::AbstractArray{Tx,Nx}, 
```

Comment:
Seems like we can unify the two methods with just converting the Cartesian indices once?

Reply:
these are very low-level methods, we should not do any allocation